velox/exec/Unnest.cpp

```diff
@@ -111,57 +112,118 @@ RowVectorPtr Unnest::getOutput() {
   }
 
   const auto size = input_->size();
-  const auto maxOutputSize = outputBatchRows();
 
   // Limit the number of input rows to keep output batch size within
-  // 'maxOutputSize' if possible. Process each input row fully. Do not break
-  // single row's output into multiple batches.
-  vector_size_t numInput = 0;
+  // 'maxOutputSize' if possible. the first and last row might not be processed
+  // fully when the output exceeds 'maxOutputSize'. The output of first and last
+  // row may be divided into multiple batches.
   vector_size_t numElements = 0;
-  for (auto row = nextInputRow_; row < size; ++row) {
-    numElements += rawMaxSizes_[row];
-    ++numInput;
-
-    if (numElements >= maxOutputSize) {
-      break;
-    }
-  }
-
+  vector_size_t partialProcessRowStart = -1;
+  auto rowRange = extractRowRange(size, numElements, partialProcessRowStart);
   if (numElements == 0) {
     // All arrays/maps are null or empty.
     input_ = nullptr;
     nextInputRow_ = 0;
     return nullptr;
   }
 
-  auto output = generateOutput(nextInputRow_, numInput, numElements);
-
-  nextInputRow_ += numInput;
+  auto output = generateOutput(rowRange, numElements);
+  if (partialProcessRowStart != -1) {
+    firstRowStart_ = partialProcessRowStart;
+    nextInputRow_ += rowRange.size - 1;
+  } else {
+    firstRowStart_ = 0;
+    nextInputRow_ += rowRange.size;
+  }
 
-  if (nextInputRow_ >= size) {
+  if (nextInputRow_ >= input_->size()) {
     input_ = nullptr;
     nextInputRow_ = 0;
   }
 
   return output;
 }
 
-void Unnest::generateRepeatedColumns(
-    vector_size_t start,
+const Unnest::RowRange Unnest::extractRowRange(
     vector_size_t size,
+    vector_size_t& numElements,
+    vector_size_t& partialProcessRowStart) {
+  VELOX_DCHECK_LT(nextInputRow_, size);
+
+  vector_size_t numInput = 0;
+  vector_size_t firstRowEnd = rawMaxSizes_[nextInputRow_];
+  vector_size_t lastRowEnd = -1;
```

Comment:
Is it better to make this variable optional to avoid hard-coding as -1 and comparing with -1?